src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -5619,14 +5621,17 @@ private Rope executeConcat(Rope left, Rope right, RubyEncoding compatibleEncodin
             return concatNode.executeConcat(left, right, compatibleEncoding.jcoding);
         }
 
-        private RubyEncoding executeCheckEncoding(Object string, Object other) {
+        private RubyEncoding executeCheckEncoding(RopeWithEncoding string, RopeWithEncoding other) {
             if (checkEncodingNode == null) {
                 CompilerDirectives.transferToInterpreterAndInvalidate();
-                checkEncodingNode = insert(CheckEncodingNode.create());
+                checkEncodingNode = insert(CheckRopeEncodingNode.create());
             }
             return checkEncodingNode.executeCheckEncoding(string, other);
         }
 
+        protected RopeWithEncoding stringToRopeWithEncoding(RubyStringLibrary strings, Object string) {
```

Comment:
I think this could make sense as a RopeWithEncoding constructor if used in multiple places.